modules/e2ee/OlmAdapter.js

```diff
@@ -470,6 +544,172 @@ export class OlmAdapter extends Listenable {
         }
+        case OLM_MESSAGE_TYPES.SAS_START: {
+            if (olmData.session) {
+                if (olmData.sas) {
+                    logger.warn(`SAS already created for participant ${pId}`);
+
+                    return;
+                }
+
+                const { uuid } = msg.data;
+
+                // Send ACCEPT.
+                const ack = {
+                    [JITSI_MEET_MUC_TYPE]: OLM_MESSAGE_TYPE,
+                    olm: {
+                        type: OLM_MESSAGE_TYPES.SAS_ACCEPT,
+                        data: {
+                            uuid
+                        }
+                    }
+                };
+
+                this._sendMessage(ack, pId);
+            } else {
+                logger.debug(`Received sas init message from ${pId} but we have no session for them!`);
+
+                this._sendError(participant, 'No session found while processing sas-init');
+            }
+            break;
+        }
+        case OLM_MESSAGE_TYPES.SAS_ACCEPT: {
+            if (olmData.session) {
+                const { commitment, uuid } = msg.data;
+                olmData.sasCommitment = commitment;
+
+                olmData.sas = new Olm.SAS()
+                const pub_key = olmData.sas.get_pubkey();
+
+                // Send KEY.
+                const ack = {
+                    [JITSI_MEET_MUC_TYPE]: OLM_MESSAGE_TYPE,
+                    olm: {
+                        type: OLM_MESSAGE_TYPES.SAS_KEY,
+                        data: {
+                            key: pub_key,
+                            isInitializer: true,
+                            uuid
+                        }
+                    }
+                };
+
+                this._sendMessage(ack, pId);
+            } else {
+                logger.debug(`Received sas accept message from ${pId} but we have no session for them!`);
+
+                this._sendError(participant, 'No session found while processing sas-accept');
+            }
+            break;
+        }
+        case OLM_MESSAGE_TYPES.SAS_KEY: {
+            if (olmData.session) {
+                if (!olmData.sas) {
+                    olmData.sas = new Olm.SAS();
+                }
+
+                if (olmData.sas.is_their_key_set()) {
+                    logger.warn('SAS already has their key!');
+
+                    return;
+                }
+
+                const { key: theirKey, isInitializer, uuid } = msg.data;
+
+                if (olmData.sasCommitment) {
+                    const olmUtil = new Olm.Utility();
+                    const commitment = olmUtil.sha256(theirKey + olmData.startContent);
+                    olmUtil.free();
+
+                    if (olmData.sasCommitment !== commitment) {
+                        throw new Error('OlmAdapter commitments mismatched');
+                    }
+                }
+
+                olmData.sas.set_their_key(theirKey);
+
+                const pub_key = olmData.sas.get_pubkey();
```

Comment:
we need to avoid sending another key from Alice if the already received Bob's key.